### PR TITLE
doc: add a document on considerations and current status of hypervisor modularization

### PR DESCRIPTION
**Pull request by junjiemao1**
To kick off the efforts on modularization, this patch introduces a document
describing the goals and general principles of modular design as well as a brief
introduction on the current status of component/module decomposition.

A detailed assignment of source files to components will be added in the future.

Tracked-On: #1842
Signed-off-by: Junjie Mao <junjie.mao@intel.com>

### CODE REVIEW

File: doc/developer-guides/modularity.rst [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+**********
+
+* Each source file shall belong to one module only. One module may consist of
+  one or multiple source files, though. A source file can be a C source, a C
+  header, an assembly file, or any other file that is built into the binary
+  image.
+* Each module shall have well-defined interfaces, including the exported
+  functions and global variables. Functions and variables that are not
+  interfaces shall be static and used inside the module only.
+* Dependencies among the modules should be acyclic. Any cyclic dependency must
```
Comment (dbkinder) [inline doc/developer-guides/modularity.rst]: This sounds like it could use more explanation.  How are cyclic dependencies resolved?

File: doc/developer-guides/modularity.rst [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  be deviated explicitly.
+* The complexity of a module shall be limited.
```
Comment (dbkinder) [inline doc/developer-guides/modularity.rst]: If this is a project requirement, how is complexity measured, and what is "too much" complexity?

File: doc/developer-guides/modularity.rst [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  inject upcall interrupts to a specific VM.
+* **Device Emulation** This component implements devices that are emulated in
+  the hypervisor itself, such as the virtual programmable interrupt controllers
+  including vPIC, vLAPIC and vIOAPIC.
+* **Passthru Management** This component manages devices that are passed-through
+  to specific VMs.
+* **Extended Device Emulation** This component implements an I/O request
+  mechanism that allow the hypervisor to forward I/O accesses from UOSes to SOS
+  for emulation.
+* **VM Management** This component manages the creation, deletion and other
```
Comment (dbkinder) [inline doc/developer-guides/modularity.rst]: The box with "vCPU Loop" is not explained.

File: doc/developer-guides/modularity.rst [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  be deviated explicitly.
+* The complexity of a module shall be limited.
+
+Architecture
+************
+
+The following figure shows the high-level components of ACRN hypervisor. ACRN
+hypervisor adopts a layered design where higher layers can invoke the interfaces
+of lower layers but not vice versa. The only exception is the invocation of
+initialization routine in the **Boot** component.
```
Comment (dbkinder) [inline doc/developer-guides/modularity.rst]: Can you explain a bit more about this exception?

File: doc/developer-guides/modularity.rst [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  dependencies among them. White-box integration tests help improve the coverage
+  of tests and identify corner cases that are hard to trigger when testing the
+  hypervisor as a whole.
+* **Configurability** Modular design makes it easy to configure certain
+  functionalities in or out. This is crucial in safety-critical scenarios
+  because absence of irrelevant code is required in both MISRA-C and functional
+  safety standards.
+* **Meet functional safety requirements** Functional safety standards explicitly
+  requires a hierarchical structure of modules in software architectural
+  design. This applies to any safety integrity level defined in IEC 61508
```
Comment (dbkinder) [inline doc/developer-guides/modularity.rst]: Spaces aren't allowed in citation references, so change to:
```
... defined in [IEC_61508-3]_ or [ISO_ 26262-6]_.
```

File: doc/developer-guides/modularity.rst [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+* **VM Management** This component manages the creation, deletion and other
+  lifecycle operations of VMs.
+* **Hypervisor Initialization** This component invokes the initialization
+  subroutines in the other components to bring up the hypervisor and start up
+  SOS in sharing mode or all the VMs in partitioning mode.
+
+References
+**********
+
+.. [IEC 61508-3] IEC 61508-3:2010, Functional safety of electrical/electronic/programmable electronic safety-related systems – Part 3: Software requirements
```
Comment (dbkinder) [inline doc/developer-guides/modularity.rst]:
```suggestion
.. [IEC_61508-3] IEC 61508-3:2010, Functional safety of electrical/electronic/programmable electronic safety-related systems – Part 3: Software requirements
```

File: doc/developer-guides/modularity.rst [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+.. [ISO 26262-6] ISO 26262-6:2011, Road vehicles - Functional safety - Part 6: Product development at the software level
```
Comment (dbkinder) [inline doc/developer-guides/modularity.rst]:
```suggestion
.. [ISO_26262-6] ISO 26262-6:2011, Road vehicles - Functional safety - Part 6: Product development at the software level
```

File: doc/developer-guides/modularity.rst [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+  requires a hierarchical structure of modules in software architectural
+  design. This applies to any safety integrity level defined in IEC 61508
+  [IEC 61508-3]_ and ISO 26262 [ISO 26262-6]_.
+
+Principles
+**********
+
+* Each source file shall belong to one module only. One module may consist of
+  one or multiple source files, though. A source file can be a C source, a C
+  header, an assembly file, or any other file that is built into the binary
```
Comment (dbkinder) [inline doc/developer-guides/modularity.rst]: what "other" files are there?

File: doc/developer-guides/modularity.rst [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+* **Testability** Modules can be integrated and tested in the reverse order of
+  dependencies among them. White-box integration tests help improve the coverage
+  of tests and identify corner cases that are hard to trigger when testing the
+  hypervisor as a whole.
+* **Configurability** Modular design makes it easy to configure certain
+  functionalities in or out. This is crucial in safety-critical scenarios
+  because absence of irrelevant code is required in both MISRA-C and functional
+  safety standards.
+* **Meet functional safety requirements** Functional safety standards explicitly
+  requires a hierarchical structure of modules in software architectural
```
Comment (dbkinder) [inline doc/developer-guides/modularity.rst]:
```suggestion
  require a hierarchical structure of modules in software architectural
```

File: doc/developer-guides/modularity.rst (hunk 1 above)
Comment (junjiemao1) [reply doc/developer-guides/modularity.rst]: I added a "Minimizing Cyclic Dependencies" section on this topic. Please help review. Thanks!

File: doc/developer-guides/modularity.rst (hunk 2 above)
Comment (junjiemao1) [reply doc/developer-guides/modularity.rst]: I added a section "Measuring Complexity" to describe how complexity is measured. The current criteria is preliminary and may be changed in the future. Please help review. Thanks!

File: doc/developer-guides/modularity.rst (hunk 3 above)
Comment (junjiemao1) [reply doc/developer-guides/modularity.rst]: Added. Thanks for pointing this out.

File: doc/developer-guides/modularity.rst (hunk 4 above)
Comment (junjiemao1) [reply doc/developer-guides/modularity.rst]: Added more description on this exception, why it is introduced and how to work it around during integration testing.

File: doc/developer-guides/modularity.rst (hunk 5 above)
Comment (junjiemao1) [reply doc/developer-guides/modularity.rst]: Updated. Thanks.

File: doc/developer-guides/modularity.rst (hunk 8 above)
Comment (junjiemao1) [reply doc/developer-guides/modularity.rst]: There is none currently. I will drop the 'other' case.

File: doc/developer-guides/modularity.rst (hunk 9 above)
Comment (junjiemao1) [reply doc/developer-guides/modularity.rst]: Updated. Thanks.

File: doc/developer-guides/modularity.rst [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+==============================
+
+Cyclic dependencies can be mostly avoided by carefully defining the boundary of
+modules. The following methods can be used when certain cyclic dependency cannot
+be resolved by design.
+
+* **Use callbacks** Callback registration and invocation help reverse dependencies
+  between two modules and break cyclic dependencies. However callbacks shall be
+  used with care due to its dynamic behavior. Send proposals or patches to the
+  acrn-dev mailing list for discussing if specific callbacks are appropriate.
```
Comment (dbkinder) [inline doc/developer-guides/modularity.rst]: Add a link to the mailing list:
```suggestion
  `acrn-dev mailing list <https://lists.projectacrn.org/g/acrn-dev>`_
   for discussing if specific callbacks are appropriate.
```

Comment (junjiemao1) [reply doc/developer-guides/modularity.rst]: Updated. Thanks.

**Comment by dongyaozu** [comment]
@dbkinder  ready to merge?
